core/ecschema-metadata/src/Context.ts

```diff
@@ -10,9 +10,42 @@ import { SchemaItem } from "./Metadata/SchemaItem";
 import { SchemaItemKey, SchemaKey } from "./SchemaKey";
 
 /**
- * @beta
+ * Holds partially-loaded schema and corresponding LoadSchema that has the promise to fully load the schema
+ * @alpha
+ */
+interface LoadingSchema {
+  schema: Schema;
+  loadSchema: LoadSchema;
+}
+/**
+ * @alpha
+ */
+export class LoadedSchemas extends Array<Schema> { }
+/**
+ * @alpha
  */
-export class SchemaMap extends Array<Schema> { }
+export class LoadingSchemas extends Array<LoadingSchema> { }
+
+/**
+ * Construct through a function that returns a promise to read the schema.
+ * When loadSchema() is called the first time, it will execute the function to actually begin the promise.
+ * When loadSchema() is called after the first time, it will just return the promise.
+ * This ensures the promise doesn't run until loadSchema() is called, and there's only one loadSchema promise per schema path.
+ * @alpha
+ */
+export class LoadSchema {

```

Comment:
You have several implementations (this one, `ReadSchemaText`, etc) of a basic pattern. Did you consider defining a generic type instead?

Reply:
I'm considering using Generics type with inheritance or using only 1 Generic class instead of LoadSchema and ReadSchemaText classes. What would you suggest me to go ahead with?

Reply:
We also have a class DelayedPromise in core\ecschema-metadata\src\DelayedPromise.ts. I'm also looking at the possibility of using that instead of LoadSchema and ReadSchemaText classes.

Reply:
DelayedPromise looks like an improvement.